src/CoreWCF.RabbitMQ.Client/src/CoreWCF/ServiceModel/Channels/RabbitMqOutputChannel.cs

```diff
@@ -0,0 +1,284 @@
+﻿using System;
+using System.IdentityModel.Selectors;
+using System.IdentityModel.Tokens;
+using System.ServiceModel;
+using System.ServiceModel.Channels;
+using System.ServiceModel.Security.Tokens;
+using System.Threading.Tasks;
+using CoreWCF.Runtime;
+using RabbitMQ.Client;
+using RabbitMQ.Client.Exceptions;
+
+namespace CoreWCF.ServiceModel.Channels
+{
+    public class RabbitMqOutputChannel : ChannelBase, IOutputChannel
+    {
+        private readonly RabbitMqChannelFactory _parent;
+        private readonly RabbitMqTransportBindingElement _transport;
+        private readonly SecurityTokenManager _securityTokenManager;
+        private readonly EndpointAddress _baseAddress;
+        private readonly Uri _via;
+        private readonly MessageEncoder _encoder;
+        private RabbitMqConnectionSettings _connectionSettings;
+        private IModel _rabbitMqClient;
+
+        internal RabbitMqOutputChannel(
+            RabbitMqChannelFactory factory,
+            RabbitMqTransportBindingElement transport,
+            SecurityTokenManager securityTokenManager,
+            MessageEncoder encoder)
+            : base(factory)
+        {
+            _parent = factory;
+            _transport = transport;
+            _securityTokenManager = securityTokenManager;
+            _encoder = encoder;
+            
+            _via = _transport.BaseAddress;
+            _baseAddress = new EndpointAddress(_via);
+        }
+
+        EndpointAddress IOutputChannel.RemoteAddress => _baseAddress;
+
+        Uri IOutputChannel.Via => _via;
+
+        protected ChannelParameterCollection ChannelParameters { get; private set; }
+
+        public override T GetProperty<T>()
+        {
+            if (typeof(T) == typeof(IOutputChannel))
+            {
+                return (T)(object)this;
+            }
+
+            if (typeof(T) == typeof(ChannelParameterCollection))
+            {
+                if (State == CommunicationState.Created)
+                {
+                    lock (ThisLock)
+                    {
+                        if (ChannelParameters == null)
+                        {
+                            ChannelParameters = new ChannelParameterCollection();
+                        }
+                    }
+                }
+                return (T)(object)ChannelParameters;
+            }
+
+            T messageEncoderProperty = _encoder.GetProperty<T>();
+            if (messageEncoderProperty != null)
+            {
+                return messageEncoderProperty;
+            }
+
+            return base.GetProperty<T>();
+        }
+
+        /// <summary>
+        /// Open the channel for use. We do not have any blocking work to perform so this is a no-op
+        /// </summary>
+        protected override void OnOpen(TimeSpan timeout)
+        {
+            OnOpenAsync(timeout).Wait();
+        }
```

Comment:
If OnOpenAsync throws, calling Wait() will throw an AggregateException with the real exception inside of it. You need to use GetAwaiter().GetResult() as that will throw the real exception.

Reply:
The other uses of Wait() in Send, Close etc also need changing.